<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px">

# Project 3: Web APIs & NLP
# Notebook 1 - Data collection

## Problem Statement 

We are a group of data scientists looking to open a cafe by making data-driven business decisions. 
In order to do this, we will first analyze posts from the 'Coffee' and 'Tea' subreddits to develop a classification model capable of segregating texts into 'Coffee' and 'Tea' categories. And then filter out the most popular flavour profiles and drinks to create a data driven menu. 

In [1]:
# import libraries

import pandas as pd 
import requests
import time
import random
import datetime

In [2]:
url = 'https://api.pushshift.io/reddit/search/submission' # using pushshift endpoints to scrap reddit submissions

In [3]:
# function for extraction of reddit submissions
def get_posts(url, subreddit, before_date):
    params = {
        'subreddit': subreddit,
        'size': 250, 
        'before': before_date
    }
    sub_reddit = []
    for i in range(60): # 15_000 posts, with a size of 250 will require 60 runs
        res = requests.get(url, params)
        if res.status_code != 200:
            print(f'Error Status Code: {res.status_code}')
        else:
            data = res.json()
            sub_reddit += data['data']
            params['before'] = sub_reddit[-1]['created_utc']
            time.sleep((random.randint(2,5)))
            print(f'Run {i} completed')
    return sub_reddit
            

In [4]:
coffee = get_posts(url, 'coffee', 1664582400)

Run 0 completed
Run 1 completed
Run 2 completed
Run 3 completed
Run 4 completed
Run 5 completed
Run 6 completed
Run 7 completed
Run 8 completed
Run 9 completed
Run 10 completed
Run 11 completed
Run 12 completed
Run 13 completed
Run 14 completed
Run 15 completed
Run 16 completed
Run 17 completed
Run 18 completed
Run 19 completed
Run 20 completed
Run 21 completed
Run 22 completed
Run 23 completed
Run 24 completed
Run 25 completed
Run 26 completed
Run 27 completed
Run 28 completed
Run 29 completed
Run 30 completed
Run 31 completed
Run 32 completed
Run 33 completed
Run 34 completed
Run 35 completed
Run 36 completed
Run 37 completed
Run 38 completed
Run 39 completed
Run 40 completed
Run 41 completed
Run 42 completed
Run 43 completed
Run 44 completed
Run 45 completed
Run 46 completed
Run 47 completed
Run 48 completed
Run 49 completed
Run 50 completed
Run 51 completed
Run 52 completed
Run 53 completed
Run 54 completed
Run 55 completed
Run 56 completed
Run 57 completed
Run 58 completed
Run 59 

In [5]:
len(coffee)

14979

In [6]:
coffee_df = pd.DataFrame(coffee)

In [7]:
coffee_df['created_utc']

0        1664581292
1        1664579910
2        1664579037
3        1664578285
4        1664577472
            ...    
14974    1641176499
14975    1641175984
14976    1641175603
14977    1641173271
14978    1641172582
Name: created_utc, Length: 14979, dtype: int64

The coffee dataset covers posts from 3<sup>rd</sup> Jan 2022 to 1<sup>st</sup> October 2022.

In [8]:
tea = get_posts(url, 'tea', 1664582400)

Run 0 completed
Run 1 completed
Run 2 completed
Run 3 completed
Run 4 completed
Run 5 completed
Run 6 completed
Run 7 completed
Run 8 completed
Run 9 completed
Run 10 completed
Run 11 completed
Run 12 completed
Run 13 completed
Run 14 completed
Run 15 completed
Run 16 completed
Run 17 completed
Run 18 completed
Run 19 completed
Run 20 completed
Run 21 completed
Run 22 completed
Run 23 completed
Run 24 completed
Run 25 completed
Run 26 completed
Run 27 completed
Run 28 completed
Run 29 completed
Run 30 completed
Run 31 completed
Run 32 completed
Run 33 completed
Run 34 completed
Run 35 completed
Run 36 completed
Run 37 completed
Run 38 completed
Run 39 completed
Run 40 completed
Run 41 completed
Run 42 completed
Run 43 completed
Run 44 completed
Run 45 completed
Run 46 completed
Run 47 completed
Run 48 completed
Run 49 completed
Run 50 completed
Run 51 completed
Run 52 completed
Run 53 completed
Run 54 completed
Run 55 completed
Run 56 completed
Run 57 completed
Run 58 completed
Run 59 

In [9]:
len(tea)

14971

In [10]:
tea_df = pd.DataFrame(tea)

In [11]:
tea_df['created_utc']

0        1664578075
1        1664575731
2        1664573221
3        1664568074
4        1664559432
            ...    
14966    1629596349
14967    1629594998
14968    1629589895
14969    1629583662
14970    1629583267
Name: created_utc, Length: 14971, dtype: int64

The tea dataset covers posts from 22<sup>nd</sup> August 2021 to 1<sup>st</sup> October 2022

In [12]:
coffee_df.to_csv('../datasets/coffee_scraped.csv', index = False)
tea_df.to_csv('../datasets/tea_scraped.csv', index = False)

To collect data for analysis of Coffee and Tea, a total of 15000 submissions from each subreddit that were submitted before 1<sup>st</sup> October 2022 0000hr were scrapped.<br>
14979 out of 15000 submissions from Coffee subreddit were collected. These submissions were posted between 3<sup>rd</sup> Jan 2022 and 1<sup>st</sup> October 2022.<br>
out of 15000 submissions from Tea subreddit were collected. These submissions were posted between 22<sup>nd</sup> August 2021 and 1<sup>st</sup> October 2022.